cmd/clusterctl/cmd/migrate.go

```diff
@@ -0,0 +1,189 @@
+	clusterv1.GroupVersion.Version,
+}
+
+var migrateCmd = &cobra.Command{
+	Use:   "migrate [SOURCE]",
+	Short: "EXPERIMENTAL: Migrate cluster.x-k8s.io resources between API versions",
+	Long: `EXPERIMENTAL: Migrate cluster.x-k8s.io resources between API versions.
+
+This command is EXPERIMENTAL and may be removed in a future release!
+
+Scope and limitations:
+- Only cluster.x-k8s.io resources are converted
+- Other CAPI API groups are passed through unchanged
+- ClusterClass patches are not migrated
+- Field order may change and comments will be removed in output
+- API version references are dropped during conversion (except ClusterClass and external
+  remediation references)
+
+Examples:
+  # Migrate from file to stdout
+  clusterctl migrate cluster.yaml
+
+  # Migrate from stdin to stdout
+  cat cluster.yaml | clusterctl migrate
+
+  # Explicitly specify target <VERSION>
+  clusterctl migrate cluster.yaml --to-version <VERSION> --output migrated-cluster.yaml`,
+
+	Args: cobra.MaximumNArgs(1),
+	RunE: func(_ *cobra.Command, args []string) error {
+		return runMigrate(args)
+	},
+}
+
+func init() {
+	migrateCmd.Flags().StringVarP(&migrateOpts.output, "output", "o", "", "Output file path (default: stdout)")
+	migrateCmd.Flags().StringVar(&migrateOpts.toVersion, "to-version", clusterv1.GroupVersion.Version, fmt.Sprintf("Target API version for migration (supported: %s)", strings.Join(supportedTargetVersions, ", ")))
+
+	RootCmd.AddCommand(migrateCmd)
+}
+
+func isSupportedTargetVersion(version string) bool {
+	for _, v := range supportedTargetVersions {
+		if v == version {
+			return true
+		}
+	}
+	return false
+}
+
+func runMigrate(args []string) error {
+	if !isSupportedTargetVersion(migrateOpts.toVersion) {
+		return errors.Errorf("invalid --to-version value %q: supported versions are %s", migrateOpts.toVersion, strings.Join(supportedTargetVersions, ", "))
+	}
+
+	fmt.Fprint(os.Stderr, "WARNING: This command is EXPERIMENTAL and may be removed in a future release!")
+
+	var input io.Reader
+	var inputName string
+
+	if len(args) == 0 {
+		input = os.Stdin
+		inputName = "stdin"
+	} else {
+		sourceFile := args[0]
+		// #nosec G304
+		// command accepts user-provided file path by design
+		file, err := os.Open(sourceFile)
+		if err != nil {
+			return errors.Wrapf(err, "failed to open input file %q", sourceFile)
+		}
+		defer file.Close()
+		input = file
+		inputName = sourceFile
+	}
+
+	// Determine output destination
+	var output io.Writer
+	var outputFile *os.File
+	var err error
+
+	if migrateOpts.output == "" {
+		output = os.Stdout
+	} else {
+		outputFile, err = os.Create(migrateOpts.output)
+		if err != nil {
+			return errors.Wrapf(err, "failed to create output file %q", migrateOpts.output)
+		}
+		defer outputFile.Close()
+		output = outputFile
+	}
+
+	// Create migration engine components
+	parser := migrate.NewYAMLParser(scheme.Scheme)
+
+	targetGV := schema.GroupVersion{
+		Group:   clusterv1.GroupVersion.Group,
```

Comment:
this is ok, but it hardcodes the group to the group of an imported version package.
i think a local constant that copies the group string might be better.
...or a proper mapping that takes the user version input and determines the group from it.